schema.go

```diff
@@ -106,3 +113,69 @@ func findBuildByBuilder(db *mgo.Database, builder string, buildnum int) (*LogKee
 	}
 	return build, nil
 }
+
+func findAndDeleteTests(db *mgo.Database) ([]bson.ObjectId, error) {
+	now := time.Now()
+	query := bson.M{
+		"started": bson.M{"$lte": now.Add(-deletePassedTestCutoff)},
+		"failed":  false,
+	}
+	var testIDs []bson.ObjectId
+	err := db.C("tests").Find(query).Sort("started").Limit(maxTests).Distinct("_id", &testIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "error finding tests")
+	}
+	filter := bson.M{
+		"_id": bson.M{"$in": testIDs},
+	}
+	_, err = db.C("tests").RemoveAll(filter)
+	if err != nil {
+		return nil, errors.Wrap(err, "error deleting tests")
+	}
+	return testIDs, nil
+}
+
+func deleteLogsByTests(db *mgo.Database, ids []bson.ObjectId) (*mgo.ChangeInfo, error) {
+	query := bson.M{
```

Comment:
You can't do a query like this because it'll collection scan. You'll either need to pass the build ID as well or find some way to do a join here. The indices on the collection are:

	{
		"v" : 1,
		"key" : {
			"_id" : 1
		},
		"name" : "_id_",
		"ns" : "buildlogs.logs"
	},
	{
		"v" : 1,
		"key" : {
			"build_id" : 1,
			"test_id" : 1,
			"seq" : 1
		},
		"name" : "build_id_1_test_id_1_seq_1",
		"ns" : "buildlogs.logs"
	},
	{
		"v" : 1,
		"key" : {
			"started" : 1
		},
		"name" : "started_1",
		"ns" : "buildlogs.logs",
		"expireAfterSeconds" : NumberLong(31536000)
	},
	{
		"v" : 1,
		"key" : {
			"build_id" : 1,
			"started" : 1
		},
		"name" : "build_id_1_started_1",
		"ns" : "buildlogs.logs",
		"background" : true
	}